src/java/org/apache/cassandra/index/sai/disk/v1/IndexSearcher.java

```diff
@@ -85,14 +85,15 @@ public static IndexSearcher open(PrimaryKeyMap.Factory primaryKeyMapFactory,
     /**
      * Search on-disk index synchronously
      *
+     * @param segment      the segment being searched
      * @param expression   to filter on disk index
      * @param keyRange     key range specific in read command, used by ANN index
      * @param queryContext to track per sstable cache and per query metrics
      * @param defer        create the iterator in a deferred state
      * @param limit        the num of rows to returned, used by ANN index
      * @return {@link RangeIterator} of sstable row ids that matches given expression
      */
-    public abstract RangeIterator<Long> search(Expression expression, AbstractBounds<PartitionPosition> keyRange, QueryContext queryContext, boolean defer, int limit) throws IOException;
+    public abstract RangeIterator<Long> search(Segment segment, Expression expression, AbstractBounds<PartitionPosition> keyRange, QueryContext queryContext, boolean defer, int limit) throws IOException;
```

Comment:
instead of pushing Segment anywhere, we could cache the global (per sstable) rowid instead of the segment rowid

Reply:
not 100% sure this is better but it seems simpler at first glance

Reply:
I originally had a comment asking if passing segment around was breaking the abstraction: https://github.com/datastax/cassandra/pull/723/commits/a5016bd030f28414fb10a11a5ef90b8854b776f2.

I think you're right that it'd be simpler. I'll check to see what I can do.

Reply:
Looks like I already had the segment's metadata object in scope, which we needed for the offset. It was trivial to remove.